crates/ide_assists/src/handlers/generate_documentation_template.rs

```diff
@@ -0,0 +1,1025 @@
+    let mut lines = Vec::new();
+    let is_unsafe = ast_func.unsafe_token().is_some();
+    let param_list = ast_func.param_list()?;
+    let ref_mut_params = ref_mut_params(&param_list);
+    let self_name: Option<String> = self_name(ast_func);
+
+    lines.push(format!("use {};", build_path(ast_func, krate_name)));
+    lines.push("".into());
+    if let Some(self_definition) = self_definition(ast_func, self_name.as_deref()) {
+        lines.push(self_definition);
+    }
+    for param_name in &ref_mut_params {
+        lines.push(format!("let mut {} = ;", param_name))
+    }
+    let function_call = function_call(ast_func, &param_list, self_name.as_deref(), is_unsafe)?;
+    let ex_helper = ExHelper { function_call, param_list, ref_mut_params, self_name };
+    Some((lines, ex_helper))
+}
+
+/// Check if the function and all its parent modules are exactly `pub`
+fn is_public(ast_func: &ast::Fn) -> bool {
+    has_pub(ast_func)
+        && ast_func
+            .syntax()
+            .ancestors()
+            .filter_map(ast::Module::cast)
+            .all(|module| has_pub(&module))
+}
+
+/// Check if visibility is exactly `pub` (not `pub(crate)` for instance)
+fn has_pub<T: HasVisibility>(item: &T) -> bool {
+    item.visibility().map(|v| v.path().is_none()).unwrap_or(false)
+}
+
+/// `None` if function without a body; some bool to guess if function can panic
+fn can_panic(ast_func: &ast::Fn) -> Option<bool> {
+    let body = ast_func.body()?.to_string();
+    let can_panic = body.contains("panic!(")
+        || body.contains("assert!(")
+        || body.contains(".unwrap()")
+        || body.contains(".expect(");
+    Some(can_panic)
+}
+
+/// Helper function to get the name that should be given to `self` arguments
+fn self_name(ast_func: &ast::Fn) -> Option<String> {
+    self_partial_type(ast_func).map(|name| to_lower_snake_case(&name))
+}
+
+/// Heper function to get the name of the type of `self`
+fn self_type(ast_func: &ast::Fn) -> Option<String> {
+    ast_func
+        .syntax()
+        .ancestors()
+        .find_map(ast::Impl::cast)
+        .and_then(|i| i.self_ty())
+        .map(|t| (t.to_string()))
+}
+
+/// Heper function to get the name of the type of `self` without generic arguments
+fn self_partial_type(ast_func: &ast::Fn) -> Option<String> {
+    let mut self_type = self_type(ast_func)?;
+    if let Some(idx) = self_type.find(|c| ['<', ' '].contains(&c)) {
+        self_type.truncate(idx);
+    }
+    Some(self_type)
+}
+
+/// Helper function to determine if the function is in a trait implementation
+fn is_in_trait_impl(ast_func: &ast::Fn) -> bool {
+    ast_func
+        .syntax()
+        .ancestors()
+        .find_map(ast::Impl::cast)
+        .and_then(|impl_| impl_.trait_())
+        .is_some()
+}
+
+/// Helper function to determine if the function definition is in a trait definition
+fn is_in_trait_def(ast_func: &ast::Fn) -> bool {
+    ast_func.syntax().ancestors().find_map(ast::Trait::cast).is_some()
+}
```

Comment:
Similar here, if we have the `hir::Fn` we can check if it comes from a trait(impl) via `fn.as_assoc_item()` then checking the container.